##PROYECTO MIN: Hecho por Javier Verde Marín
#ANÁLISIS SOBRE EL COVID-19 MEDIANTE TWITTER

La finalidad de este trabajo es observar los problemas o limitaciones que pueden surgir a la hora de intentar hacer un análisis utilizando los datos de Twitter como herramienta, mientras que se elabora un estudio centrado en el Covid-19.

Este proyecto se divide en dos partes:
- La primera primera será un análisis de los datos que se pueden extraer de los json de Twitter aplicados al Covid-19.
- La segunda parte será un sentiment analysis de los tweets relacionados con el Covid-19.

Los datos han sido recopilados cargando tweets a modo de stream y luego almacenándolos para su futuro uso y evaluación, por lo que el uso de los comandos de borrado y cargado de tweets producirá el borrado de los datos (en formato json) usados para el análisis siendo reemplazados por los nuevos datos. Esto podría resultar en que los nuevos datos recopilados puedan no coincidir con los comentarios realizados a lo largo de este proyecto.

En el curso del proyecto se extraerán 1000 tweets, que utilizaremos para analizar y sacar las conclusiones de la mayoría de las pruebas y otros 1000 tweets exclusivamente en inglés para poder realizar correctamente el sentiment analysis, ya que no clasifica los textos de otros idiomas correctamente y el proceso de traducción es demasiado costoso. Todos los tweets que se utilizarán como prueba tratarán sobre algún tema relacionado al Covid-19 pues pasarán por un filtro para no salirnos del estudio.

Para extraer la información de los tweets se usarán los campos del formato json en el que Twitter los almacena. Cada campo hace referencia a algún metadato o parte que compone el tweet, como la fecha, el idioma, el texto, etc. 
El propio texto del tweet será de especial importancia en la segunda parte del proyecto, porque es el mensaje a transmitir lo que queremos clasificar en positivo o negativo para su sentiment analysis.

Las herramientas que se van a utilizar son:
- **Tweepy** (para extraer los tweets a través de la API de Twitter).
- **Textblob** (analizador por defecto y analizador Naive Bayes para realizar el sentiment analysis).

Finalmente se redactarán las conclusiones a las que se ha llegado tras el análisis y estudio del caso.

##Parte 1: Análisis y extracción de los datos de Twitter sobre el Covid-19

En esta parte vamos a ver cómo podemos utilizar los distintos tipos de datos que Twitter nos proporciona en cada tweet para extraer información relacionada al Covid-19 y comprobar si esta es suficiente para llegar a conclusiones concretas.

Primero vamos a instalar tweepy (https://www.tweepy.org/) y otras bibliotecas para poder interactuar con la API de Twitter:

In [0]:
%pip install tweepy
%pip install -U textblob

Python interpreter will be restarted.
Collecting tweepy
 Downloading tweepy-3.10.0-py2.py3-none-any.whl (30 kB)
Requirement already satisfied: requests[socks]>=2.11.1 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-880d10ce-2596-4d5c-add8-2349a88bc915/lib/python3.8/site-packages (from tweepy) (2.24.0)
Requirement already satisfied: requests-oauthlib>=0.7.0 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-880d10ce-2596-4d5c-add8-2349a88bc915/lib/python3.8/site-packages (from tweepy) (1.3.0)
Requirement already satisfied: six>=1.10.0 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-880d10ce-2596-4d5c-add8-2349a88bc915/lib/python3.8/site-packages (from tweepy) (1.15.0)
Requirement already satisfied: certifi>=2017.4.17 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-880d10ce-2596-4d5c-add8-2349a88bc915/lib/python3.8/site-packages (from requests[socks]>=2.11.1->tweepy) (2020.12.5)
Requirement already satisfied: chardet<4,>=3.0.2 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-880d10ce-2596-4d5c-add8-2349a88bc915/lib/python3.8/site-packages (from requests[socks]>=2.11.1->tweepy) (3.0.4)
Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-880d10ce-2596-4d5c-add8-2349a88bc915/lib/python3.8/site-packages (from requests[socks]>=2.11.1->tweepy) (1.25.11)
Requirement already satisfied: idna<3,>=2.5 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-880d10ce-2596-4d5c-add8-2349a88bc915/lib/python3.8/site-packages (from requests[socks]>=2.11.1->tweepy) (2.10)
Requirement already satisfied: PySocks!=1.5.7,>=1.5.6; extra == "socks" in /local_disk0/.ephemeral_nfs/envs/pythonEnv-880d10ce-2596-4d5c-add8-2349a88bc915/lib/python3.8/site-packages (from requests[socks]>=2.11.1->tweepy) (1.7.1)
Requirement already satisfied: oauthlib>=3.0.0 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-880d10ce-2596-4d5c-add8-2349a88bc915/lib/python3.8/site-packages (from requests-oauthlib>=0.7.0->tweepy) (3.1.0)
Installing collected packages: tweepy
Successfully installed tweepy-3.10.0
Python interpreter will be restarted.
Python interpreter will be restarted.
Collecting textblob
 Downloading textblob-0.15.3-py2.py3-none-any.whl (636 kB)
Requirement already satisfied, skipping upgrade: nltk>=3.1 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-880d10ce-2596-4d5c-add8-2349a88bc915/lib/python3.8/site-packages (from textblob) (3.5)
Requirement already satisfied, skipping upgrade: regex in /local_disk0/.ephemeral_nfs/envs/pythonEnv-880d10ce-2596-4d5c-add8-2349a88bc915/lib/python3.8/site-packages (from nltk>=3.1->textblob) (2020.10.15)
Requirement already satisfied, skipping upgrade: click in /local_disk0/.ephemeral_nfs/envs/pythonEnv-880d10ce-2596-4d5c-add8-2349a88bc915/lib/python3.8/site-packages (from nltk>=3.1->textblob) (7.1.2)
Requirement already satisfied, skipping upgrade: tqdm in /local_disk0/.ephemeral_nfs/envs/pythonEnv-880d10ce-2596-4d5c-add8-2349a88bc915/lib/python3.8/site-packages (from nltk>=3.1->textblob) (4.50.2)
Requirement already satisfied, skipping upgrade: joblib in /local_disk0/.ephemeral_nfs/envs/pythonEnv-880d10ce-2596-4d5c-add8-2349a88bc915/lib/python3.8/site-packages (from nltk>=3.1->textblob) (0.17.0)
Installing collected packages: textblob
Successfully installed textblob-0.15.3
Python interpreter will be restarted.

In [0]:
import tweepy 
from tweepy import OAuthHandler
from tweepy import Stream 
from tweepy.streaming import StreamListener
import socket 
import json 

Introducimos los datos de la API de Twitter (Los copiamos del panel de Twitter Developer al crear la app).

In [0]:
access_token = "*****"
access_token_secret = "******"
API_key = "*******"
API_secret_key = "*********"

Necesitamos poder enviar y filtrar los datos.
* Guardamos la autenticación (las claves de la API) : https://docs.tweepy.org/en/latest/auth_tutorial.html

In [0]:
auth = OAuthHandler(API_key, API_secret_key)
auth.set_access_token(access_token, access_token_secret)

Necesitamos crear un listener que pueda recibir los tweets
* Los tweets se tratarán a modo de un json: https://docs.tweepy.org/en/latest/streaming_how_to.html (Paso 1)

* Si la app de Twitter está enviando demasiadas peticiones y se produce error, paramos el streaming: https://developer.twitter.com/ja/docs/basics/response-codes

* Guardamos cada tweet en un archivo, cuyo número de identificación se va incrementando.

In [0]:
class MyStreamListener(tweepy.StreamListener):
    def __init__(self,api=None):
        super(MyStreamListener,self).__init__()
        self.num_tweets=0
    def on_status(self,status):
        tweet= status._json
        aux = json.dumps(tweet)
        print(aux)
        dbutils.fs.put("/FileStore/tables/tweets/tweet-"+str(self.num_tweets)+".json", aux)
        self.num_tweets+=1;
        if self.num_tweets<1000 :
            return True
        else:
            return False
    def on_error(self, status):
        print(status)
        if status == 420: return False
        return True

En caso de que necesitemos eliminar los tweets almacenados, por ejemplo para otro muestreo usaremos este comando:

**No activarlo** si se desea mantener las mismas muestras pues eso borrará los datos.

In [0]:
dbutils.fs.rm('/FileStore/tables/tweets/', True)

Out[33]: True

**Activarlo solo** si se desean muestras nuevas tras borrar las muestras previas.

Creamos el stream y filtramos:
* Creamos el stream para poder recibir los tweets: https://docs.tweepy.org/en/latest/streaming_how_to.html (Paso 2)
* Introducimos las condiciones de filtrado (en nuestro caso que contengan los distintos referentes del covid). (Paso 3)
* Se irán guardando a modo de ficheros en el directorio /tweets.

In [0]:
stream = Stream(auth, MyStreamListener())
stream.filter(track=['covid','covid-19','covid19','coronavirus'])

Comprobamos que se hayan guardado los tweets:

In [0]:
%fs ls /FileStore/tables/tweets/

path name size dbfs:/FileStore/tables/tweets/tweet-0.json tweet-0.json 5324 dbfs:/FileStore/tables/tweets/tweet-1.json tweet-1.json 4842 dbfs:/FileStore/tables/tweets/tweet-10.json tweet-10.json 5385 dbfs:/FileStore/tables/tweets/tweet-100.json tweet-100.json 5494 dbfs:/FileStore/tables/tweets/tweet-101.json tweet-101.json 6004 dbfs:/FileStore/tables/tweets/tweet-102.json tweet-102.json 5399 dbfs:/FileStore/tables/tweets/tweet-103.json tweet-103.json 4919 dbfs:/FileStore/tables/tweets/tweet-104.json tweet-104.json 2926 dbfs:/FileStore/tables/tweets/tweet-105.json tweet-105.json 6014 dbfs:/FileStore/tables/tweets/tweet-106.json tweet-106.json 5242 dbfs:/FileStore/tables/tweets/tweet-107.json tweet-107.json 5747 dbfs:/FileStore/tables/tweets/tweet-108.json tweet-108.json 6664 dbfs:/FileStore/tables/tweets/tweet-109.json tweet-109.json 5355 dbfs:/FileStore/tables/tweets/tweet-11.json tweet-11.json 5046 dbfs:/FileStore/tables/tweets/tweet-110.json tweet-110.json 8999 dbfs:/FileStore/tables/tweets/tweet-111.json tweet-111.json 5776 dbfs:/FileStore/tables/tweets/tweet-112.json tweet-112.json 5182 dbfs:/FileStore/tables/tweets/tweet-113.json tweet-113.json 2315 dbfs:/FileStore/tables/tweets/tweet-114.json tweet-114.json 6420 dbfs:/FileStore/tables/tweets/tweet-115.json tweet-115.json 4710 dbfs:/FileStore/tables/tweets/tweet-116.json tweet-116.json 5756 dbfs:/FileStore/tables/tweets/tweet-117.json tweet-117.json 7594 dbfs:/FileStore/tables/tweets/tweet-118.json tweet-118.json 11643 dbfs:/FileStore/tables/tweets/tweet-119.json tweet-119.json 6197 dbfs:/FileStore/tables/tweets/tweet-12.json tweet-12.json 12263 dbfs:/FileStore/tables/tweets/tweet-120.json tweet-120.json 5309 dbfs:/FileStore/tables/tweets/tweet-121.json tweet-121.json 2936 dbfs:/FileStore/tables/tweets/tweet-122.json tweet-122.json 5906 dbfs:/FileStore/tables/tweets/tweet-123.json tweet-123.json 8513 dbfs:/FileStore/tables/tweets/tweet-124.json tweet-124.json 4810 dbfs:/FileStore/tables/tweets/tweet-125.json tweet-125.json 5945 dbfs:/FileStore/tables/tweets/tweet-126.json tweet-126.json 3126 dbfs:/FileStore/tables/tweets/tweet-127.json tweet-127.json 9364 dbfs:/FileStore/tables/tweets/tweet-128.json tweet-128.json 5512 dbfs:/FileStore/tables/tweets/tweet-129.json tweet-129.json 6234 dbfs:/FileStore/tables/tweets/tweet-13.json tweet-13.json 4992 dbfs:/FileStore/tables/tweets/tweet-130.json tweet-130.json 5712 dbfs:/FileStore/tables/tweets/tweet-131.json tweet-131.json 12169 dbfs:/FileStore/tables/tweets/tweet-132.json tweet-132.json 7331 dbfs:/FileStore/tables/tweets/tweet-133.json tweet-133.json 6064 dbfs:/FileStore/tables/tweets/tweet-134.json tweet-134.json 2366 dbfs:/FileStore/tables/tweets/tweet-135.json tweet-135.json 5589 dbfs:/FileStore/tables/tweets/tweet-136.json tweet-136.json 6458 dbfs:/FileStore/tables/tweets/tweet-137.json tweet-137.json 5621 dbfs:/FileStore/tables/tweets/tweet-138.json tweet-138.json 8243 dbfs:/FileStore/tables/tweets/tweet-139.json tweet-139.json 5307 dbfs:/FileStore/tables/tweets/tweet-14.json tweet-14.json 7547 dbfs:/FileStore/tables/tweets/tweet-140.json tweet-140.json 5712 dbfs:/FileStore/tables/tweets/tweet-141.json tweet-141.json 7459 dbfs:/FileStore/tables/tweets/tweet-142.json tweet-142.json 4933 dbfs:/FileStore/tables/tweets/tweet-143.json tweet-143.json 6485 dbfs:/FileStore/tables/tweets/tweet-144.json tweet-144.json 12360 dbfs:/FileStore/tables/tweets/tweet-145.json tweet-145.json 11522 dbfs:/FileStore/tables/tweets/tweet-146.json tweet-146.json 5912 dbfs:/FileStore/tables/tweets/tweet-147.json tweet-147.json 18524 dbfs:/FileStore/tables/tweets/tweet-148.json tweet-148.json 4935 dbfs:/FileStore/tables/tweets/tweet-149.json tweet-149.json 4756 dbfs:/FileStore/tables/tweets/tweet-15.json tweet-15.json 9793 dbfs:/FileStore/tables/tweets/tweet-150.json tweet-150.json 5576 dbfs:/FileStore/tables/tweets/tweet-151.json tweet-151.json 5722 dbfs:/FileStore/tables/tweets/tweet-152.json tweet-152.json 4776 dbfs:/FileStore/tables/tweets/twee

Es importante conocer el formato del json de Twitter y sus componentes para poder saber de dónde y cómo extraer los datos:
  * Data dictionary de Twitter: https://developer.twitter.com/en/docs/twitter-api/v1/data-dictionary/overview

Definimos el dataframe que representará al json (importante tener la opción **multiline** activa):

In [0]:
tweetInputDF = (
  spark
    .read
    .option("multiline", True)
    .json("/FileStore/tables/tweets/")
)

Podemos observar el contenido de los tweets mediante collect():

Van a ser necesarios el **texto** del tweet (se utilizará más adelante en la parte 2), el **idioma** detectado y los **json anidados** que más nos interesan, como: **place** (información de geolocalización), **user** (información de la cuenta que publica el tweet, incluyendo lugar introducido por el usuario) y **entities** (que contiene datos como los hashtags).

In [0]:
tweetCountsDF = (
  tweetInputDF
    .groupBy(
       tweetInputDF.lang,
       tweetInputDF.text,
       tweetInputDF.place,
       tweetInputDF.user,
       tweetInputDF.entities
    )
    .count()
)
tweetCountsDF.cache()


Out[14]: DataFrame[lang: string, text: string, place: struct<bounding_box:struct<coordinates:array<array<array<double>>>,type:string>,country:string,country_code:string,full_name:string,id:string,name:string,place_type:string,url:string>, user: struct<contributors_enabled:boolean,created_at:string,default_profile:boolean,default_profile_image:boolean,description:string,favourites_count:bigint,follow_request_sent:string,followers_count:bigint,following:string,friends_count:bigint,geo_enabled:boolean,id:bigint,id_str:string,is_translator:boolean,lang:string,listed_count:bigint,location:string,name:string,notifications:string,profile_background_color:string,profile_background_image_url:string,profile_background_image_url_https:string,profile_background_tile:boolean,profile_banner_url:string,profile_image_url:string,profile_image_url_https:string,profile_link_color:string,profile_sidebar_border_color:string,profile_sidebar_fill_color:string,profile_text_color:string,profile_use_background_image:boolean,protected:boolean,screen_name:string,statuses_count:bigint,time_zone:string,translator_type:string,url:string,utc_offset:string,verified:boolean>, entities: struct<hashtags:array<struct<indices:array<bigint>,text:string>>,media:array<struct<additional_media_info:struct<monetizable:boolean>,display_url:string,expanded_url:string,id:bigint,id_str:string,indices:array<bigint>,media_url:string,media_url_https:string,sizes:struct<large:struct<h:bigint,resize:string,w:bigint>,medium:struct<h:bigint,resize:string,w:bigint>,small:struct<h:bigint,resize:string,w:bigint>,thumb:struct<h:bigint,resize:string,w:bigint>>,source_status_id:bigint,source_status_id_str:string,source_user_id:bigint,source_user_id_str:string,type:string,url:string>>,symbols:array<string>,urls:array<struct<display_url:string,expanded_url:string,indices:array<bigint>,url:string>>,user_mentions:array<struct<id:bigint,id_str:string,indices:array<bigint>,name:string,screen_name:string>>>, count: bigint]

Creamos una tabla para poder realizar las siguients consultas SQL:
* Calcular los tweets relacionados por región usando geolocalización.
* Calcular los tweets relacionados por región usando el lugar introducido por el usuario.
* Calcular los tweets relacionados dividiendo los tweets en idiomas.
* Contar el número de apariciones de los hashtags utilizados en estos tweets.

In [0]:
tweetCountsDF.createOrReplaceTempView("tweets19")

#####Calcular los tweets relacionados por región usando geolocalización

In [0]:
%sql select place.country_code, sum(count) as total_count from tweets19 group by place.country_code

country_code total_count null 993 US 3 ES 1 BR 1 AR 1 GB 1

En este caso se puede observar que apenas hay valores más allá de **null** en el campo de place (json anidado dentro del json del tweet):
* Place solo tiene contenido si el usuario tiene activada la geolocalización.
* Por defecto la geolocalización del tweet no está activada.
* Más sobre el json place: https://developer.twitter.com/en/docs/twitter-api/v1/data-dictionary/object-model/geo

De esta forma podemos deducir que los datos de región provenientes de la geolocalización de Twitter son demasiado escasos y no se puede sacar una deducción concreta mediante estos.

* Con los resultados obtenidos, se podría llegar a la conclusión de que es posible que haya un mayor número de tweets relacionados procedentes de los países que están sufriendo una peor situación debido al Covid-19, en comparación con otros. 
* Los valores no nulos representan a Estados Unidos, España, Reino Unido, Brasil y Argentina (países muy afectados según Statista):
https://es.statista.com/estadisticas/1091192/paises-afectados-por-el-coronavirus-de-wuhan-segun-los-casos-confirmados/

No obstante, solo sería una deducción ya que tal escasez de datos (993 valores nulos de 1000 muestras) no sería concluyente.

#####Calcular los tweets relacionados por región usando el lugar introducido por el usuario

In [0]:
%sql select user.location, sum(count) as total_count from tweets19 group by user.location

location total_count Some Where Over the rainbow🌈 1 null 396 Montréal, Québec 3 Hyderabad 1 NANAIMO BC CANADA 1 Durango, México 1 Lothlórien - Terra Média 1 Ontario, Canada 2 الرياض, المملكة العربية السعود 2 Cumaná, Edo. Sucre - Venezuela 2 Naarm (Melb) & Cadi (Sydney) 1 Philadelphia, PA 1 Rio de Janeiro, Brasil 5 Internet 2 Rio Grande do Sul - Brasil 1 Chetoeslovaquia 1 London, Ontario, Canada 1 Portland, ME 1 Porto Alegre - Brasil 1 Caucaia, Brasil 1 Barcelona, España 1 De los matorrales 📍 1 Tánger - Tetuán 1 Ironforge 1 Fort McMurray, Alberta 1 Vancouver Island 3 In The Present 1 emi éssi 1 City of Flowers 1 bs.as 1 lutetia parisiorum 1 Sao Paulo, SP, Brazil 1 日本 2 Mississauga-Treaty 13a, ON🇨🇦 1 Lorraine 1 San Francisco, Zulia 2 Selfish Tory England 1 Ciudad Guayana, Venezuela 1 Iowa, USA 1 Australia 2 Hogwarts 1 Canada 4 Cape Town, South Africa 1 Cyber Citizen (Netizen) 1 Nova Iorque, USA 3 The Blue Marble 1 California, USA 2 Orlando, FL 2 Venezuela 16 01 21 4 Ecaterimburgo, Rússia 2 United States 8 Campos dos Goytacazes, Brasil 1 Gravatá - Pernambuco, Brasil 1 Dreux, France 1 Somewhere, flying over the 🌏 1 Bronx, New York 1 Perdido 1 Ciudad Autónoma de Buenos Aire 3 Hillingdon, London 1 Instagram: @Oticastro 1 Brasil 10 Capital Federal 1 Toronto, Ontario 2 Auf Twitter dot com 1 Cachoeira do Sul, Brasil 1 Indiana, USA 1 Union Gap, WA 1 Bogotá, D.C., Colombia 1 inside my head, mostly 2 São Paulo 1 Limerick, Ireland 1 Siempre Venezuela 1 1 Flyover Country 1 venezuela 2 Estados Unidos 2 Paraguay 3 Lucknow, India 1 Dallas, TX 1 San Diego, CA 3 Curitiba, Brasil 2 Tír na nÓg 1 icebergfinanza.finanza.com 1 Ciudad Juárez, Chihuahua 1 France Tarn et Garonne 1 Toronto 3 Porto Alegre, Brasil 2 São Paulo, Brasil. 1 Ohio 1 Las Vegas, NV 1 B_L_M 1 tempe az she/her 1 New York, NY 3 Florianópolis, Brasil 1 Belo Horizonte, Brasil 3 Stockholm, Warszawa 1 Los Angeles County 1 Broken Arrow, OK 1 Arizona, USA 1 Ouro Fino, Brasil 1 Punto Fijo, Edo Falcón 1 Uberaba - MG, Brasil 1 Georgia, USA 1 the Bay Area, CA 1 São Paulo, Brasil 7 Eora Land (Sydney) Australia 1 Buenos Aires, Argentina 4 New York 1 Toscana, Italia 1 Los Rios 1 Arizona 1 São João da Boa Vista, Brasil 1 Banana Republic USA 1 Oep a bakkes 1 Guadalajara, Jalisco 2 Los Angeles, CA 3 Viamão, Brasil 1 พญาไท, กรุงเทพมหานคร 3 Washington, DC 2 Raleigh, NC 1 México 1 Boulder, CO 1 Texas, USA 2 London, England 1 Yorkshire 1 The Lookout (In the HTC) 1 NOVO BRASIL/ DESINFORMAÇÃO NÃO 1 Brazil 2 Earth, The New bits o' England 1 welcome in the world 1 Ocean City, Maryland 1 ituzaingo . 1 Auburn, AL 1 Chapada Diamantina - BAHIA 1 Where the money reside..... 1 Valencia, Comunidad Valenciana 1 Star Prairie, WI 2 Venezuela 9 onde a inveja n chega 1 Emirates Stadium 1 Republic of Texas 1 louis... 1 Rio de Janeiro 2 Malaga, Spain 1 Montego Bay 1 Tkaronto/Toronto 1 Cape Town 1 Sao Paulo, Brazil 1 Firenze, Toscana 1 Hampshire, England 1 The Bebop, #GeeksResist HQ 1 Poznań,Hemel Polska/Anglia 1 The Other Side of Down 1 Chester, PA 1 Houston, TX 3 Glendale Az 1 Rajasthan, India 1 Santa Marta, Colombia 2 C. Autónoma de Buenos Aires 1 . 1 Somewhere in Illinois 1 England 2 Moncalieri (TO) 1 she/they ~ minor 1 Rio Grande do Sul, Brasil 2 L.A. & Coachella Valley 1 Detroit, MI 1 Vantablack Tattoo Gallery 1 Misson, TX 1 ÜT: -8.209973,-34.923087 1 UK/Algeria/Exile 1 Valleseco, Canary Islands 1 London, Edinburgh, Ercan, 1 Cuautitlán Izcalli, México 1 Sevilla, España 1 Ma. 1 Curitiba, Brazil 1 Madrid 1 Hants.orig.from London 1 Foggia 1 Helsinki, Suomi-Finland 1 061 1 Cienfuegos,Cuba 1 En la magia del sur Chile 1 Duns 1 Monterrey, Nuevo León 1 Xalapa-Enriquez, Ver. 1 Guaynabo, Puerto Rico 1 Arrakis

 1 my bedroom 1 Suffolk, VA 1 Ville-Émarde, Montréal 1 Coruña (Spain) 1 out of my skin 1 Deutschland, Stuttgart 1 New York, NY Atlanta, GA 1 𝗈𝗇𝗅𝗒 𝖾𝗑𝗈𝗅 ㋛︎ 1 Vancouver, BC 2 En el exilio 1 UCL, South Cloisters 1 San Fernando de APURE 1 Wisconsin 1 Near Parry Sound, Ontario atop tall red pine. 1 Córdoba - Argentina 1 Land 

Por otro lado, si lo que usamos es el valor location que se encuentra en el json de user, recibimos una gran cantidad de datos en comparación con el caso anterior.

Sin embargo, vuelven a surgir problemas:
* Las personas pueden introducir el lugar que quieran: Internet, Pokemon, Some Where Over the rainbow🌈...por nombrar algunos ejemplos.
* No hay una uniformidad de criterio, se puede introducir tanto países como códigos de país, calles, ciudades....

Es por esto que no se puede confiar en los datos proporcionados, pues muchos son lugares inventados.
En el resto de casos sí podría existir la localización que los usuarios indiquen, pero darse el caso de que el dato sea falso. Además, si lo que el usuario ha introducido es una ciudad o una dirección, no se reconocerá el lugar como parte de un país, sino que permanecerá como una nueva localización.

#####Calcular los tweets relacionados dividiendo los tweets en idiomas

In [0]:
%sql select lang, sum(count) as total_count from tweets19 group by lang

lang total_count en 393 pt 185 es 290 th 4 ar 3 zh 4 tl 3 und 39 fr 29 ja 2 it 15 de 6 tr 11 ca 3 sv 1 pl 4 nl 2 cy 1 fi 1 lv 1 ko 1 in 1 ro 1

Debido a la falta de información (o al menos información útil) en los dos casos anteriores, en vez de analizar por gelocalización y localización, vamos a analizar los tweets mediante los idiomas empleados.
Twitter identifica el idioma evaluando los tweets, el cuál podemos obtener en el campo **lang** del json.
Más sobre cómo Twitter evalúa los tweets para identificar el idioma: https://blog.twitter.com/engineering/en_us/a/2015/evaluating-language-identification-performance.html

Se devuelve un código de idioma que sigue las normas de la ISO-639-1: https://www.andiamo.co.uk/resources/iso-language-codes/ (un poco atrasado ya que en la actualidad se usa la ISO-639-5 que se basa en códigos de 3 letras.)

En caso de que el tweet solo contenga símbolos, nombres, canciones, etc, no será información suficiente y por tanto se considerará como **und** (indefinido/no catalogado). En este caso, de 1000 tweets, solo se han encontrado 39 indefinidos; lo cuál contrasta con los 993 nulls del análisis mediante geolocalización.

Se puede observar que los idiomas más utilizados han sido:
* Inglés **(39%)**
* Español **(29%)**
* Portugués **(19%)**

Debido a que ahora la cantidad de datos contretos en la muestra es considerable, podemos encontrar que efectivamente hay una correlación al menos entre los idiomas en los que los tweets están escritos y las zonas más afectadas.
- USA y Reino Unido siguen en el ranking de países con más casos.
- Muchos países hispanohablantes has sido afectados gravemente por la pandemia.
- En Brasil se habla portugués, y es también de los países más afectados.

El único problema es que no podemos saber cuál es la región concreta de la que provienen los tweets, pero aporta sin lugar a dudas mucha más información que las consultas previas. Además nos puede mostrar qué comunidades presentan más preocupaciones o se expresan más con respecto al Covid-19.

#####Contar el número de apariciones de los hashtags utilizados en estos tweets

El json de los tweets además contiene otro json anidado llamado **entities**, que proporciona datos como los hashtags, urls, símbolos utilizados, etc...
Para más información: https://developer.twitter.com/en/docs/twitter-api/v1/data-dictionary/object-model/entities

Por tanto, ésto nos puede ser útil a la hora de buscar los hashtags más populares (teniendo en cuenta que partimos de las muestras) y otra información variada.

En este caso vamos a probar su uso calculando las veces que se repite cada hashtag en los tweets tomados como muestra, que nos puede ayudar a aportar un pensamiento general de los usuarios acerca del tema.

Vamos a aprovecharnos de las utilizades de los RDD, para extraer los diferentes hashtags de los tweets e irlos almacenando en un map para poder contarlos.
Luego podemos convertir el RDD en un DF para poder transformarlo fácilmente en una tabla y usar las consultas SQL.

**Se parasarán todos los hashtags a minúscula ya que el mensaje a transmitir es el mismo.**

In [0]:
hashtagRDD =(tweetCountsDF.rdd
              .flatMap(lambda tweet: [hashtag['text'].lower() for hashtag in tweet['entities']['hashtags']])
              .map(lambda ht: ("#"+ht, 1))
              .reduceByKey(lambda a, b: a + b)
            )
hashtagDF = spark.createDataFrame(hashtagRDD).toDF("Hashtag", "count")
hashtagDF.createOrReplaceTempView("hashtags19")

In [0]:
%sql select Hashtag, count from hashtags19 order by count desc

Hashtag count #covid19 60 #6abr 20 #cuidémonosdelvirus 15 #coronavirus 6 #urgente 5 #cuba 5 #covid 5 #apure 4 #maduroguerralive 4 #longcovid 3 #corona 3 #pandemia 3 #liarfromtheshire 3 #elcovidnoesjuegocuidate 3 #esnoticia 3 #escontodxsadentro 2 #共产党 2 #sarscov2 2 #apureesnuestro 2 #爆料革命 2 #ccp 2 #pandemic 2 #新中国联邦 2 #sidor 2 #pdvsa 2 #actualización 2 #cuarentenamg 1 #كورونا 1 #leadership 1 #longcovidkids 1 #odia 1 #gotoecarri 1 #zerocovid 1 #covid19ab 1 #womeninbusiness 1 #cundinamarca 1 #broward 1 #gouvernement 1 #essaporramata 1 #fotos 1 #audit 1 #pointsituation 1 #like 1 #turkey 1 #الضفدع_بره_الزمالك 1 #vacunarseessalud 1 #quilombola 1 #nomasksinclass 1 #michaelyeadon 1 #โควิด19 1 #umasónação 1 #イベルメクチン用量 1 #اطلقوا_سجناء_البحرين 1 #endirect 1 #cubaporlasalud 1 #sinopharm 1 #foxnews 1 #gremio 1 #argentina 1 #onpoli 1 #ahora 1 #notimippci 1 #realestate 1 #ayuso 1 #live 1 #covidorigin 1 #イベルメクチン服用量 1 #cancelboardexams2021 1 #ecolealamaison 1 #últimahora 1 #entrepreneurmindset 1 #deltaamacuro 1 #infirmière 1 #losangelescounty 1 #americanjobsplan 1 #lomásvistodeldía 1 #vaccins 1 #enfotos 1 #middlesex 1 #american 1 #breaking 1 #katrinakaif 1 #durango 1 #drosten 1 #covidontario 1 #alerta 1 #شهر_رمضان 1 #successful 1 #getvaccinated 1 #exclusivo 1 #ivermectina 1 #hardwork 1 #vaccin 1 #dsbn 1 #morrisonrecession 1 #galicia 1 #buenosaires 1 #patheticresponse 1 #contigochv 1 #اعلان_موعد_ليله_القدر_1442 1 #fuengirola 1 #covidー19 1 #thefive 1 #quiéndijoqueestarde 1 #radiofórmulajuárez 1 #newdeathparty 1 #gnbdttocapital 1 #businesscoach 1 #vacinaparatodos 1 #taqueuspariu 1 #ethiopia 1 #actualidadcf 1 #vaccination 1 #canucks 1 #auxilio 1 #regionales 1 #sabelo 1 #noticias 1 #adunata 1 #bollywood 1 #biden 1 #america 1 #ldnont 1 #believe 1 #insurance 1 #bolívar 1 #yeşilyurt 1 #education 1 #paris 1 #vigo 1 #happs 1 #news 1 #editorial 1 #periscope 1 #cdmx 1 #dimartedi 1 #iifa 1 #ai 1 #president 1 #5abr 1 #ig 1

Podemos observar que estos datos son mucho más descriptivos, variados y abundan más que las dos primeras pruebas. 
No solo están los hashtags esperados como covid, كورون (corona), coronavirus o covid19, también encontramos hashtags que muestran la preocupación de la gente acerca de este tema:
  * cuidémonosdelvirus
  * elcovidnoesunjuegocuidate
  * zerocovid
  * ivermectina (medicamento que se está desaconsejando usar para tratar la enfermedad) y イベルメクチン用量 (
Dosis de ivermectina)
  * vaccine
  * cancelboardexams2021 (se está pidiendo que cierren debido al auge del virus)
  * cubaporlasalud
  * entre otros
 
  
Esta información puede resultar muy valiosa pues resume las distintas opiniones, puntos de vista y preocupaciones de los usuarios además de ser mucho más tangible al igual que el análisis que utiliza lang, en comparación con los otros datos que hemos probado antes (en parte por las medidas de privacidad que Twitter deja a tu disposición como por ejemplo: tener activada o no la geolocalización).
- En caso de que se desee echar un vistazo:
    * https://twitter.com/en/privacy
    * https://help.twitter.com/en/safety-and-security#ads-and-data-privacy
      
No obstante, si de algún lugar podemos extraer la mayor cantidad de información, es en el campo **texto** (que no está en ningún json anidado) ya que es lo que contiene el mensaje, es decir, lo que el usuario desea transmitir. Por ello, pasaremos a la segunda parte de este proyecto.

##Parte 2: Twitter Sentiment Analysis orientado al Covid-19

En esta parte del proyecto, vamos a analizar los sentimientos de los usuarios a partir de los tweets, utilizando el campo text del json, que es el que contiene el texto del tweet. De esta manera, obtendremos una interesante cantidad de información que nos ayudará a comprender las opiniones y sentimientos y hacia dónde se inclina la balanza al hablar del Covid-19.

Ya se ha instalado con anterioridad textblob (https://textblob.readthedocs.io/en/dev/) para poder procesar **lenguaje natural (NLP)**

In [0]:
import textblob
from textblob import TextBlob
from textblob import Blobber

#####Twitter Sentiment Analysis utilizando la opción por defecto

Esta opción usa una biblioteca de patrones: https://github.com/clips/pattern/blob/d25511f9ca7ed9356b801d8663b8b5168464e68f/pattern/text/en/wordnet/__init__.py#L490)

Vamos a convertir el dataframe utilizado anteriormente para las estadísticas previas en un RDD para poder aprovechar de nuevo sus utilidades. Se crea un map que contiene el tweet, la polaridad (el sentimiento negativo o positivo) y la subjetividad (si se cierne a opiniones o hechos).

**Se presenta un problema importante:** los tweets que no están en inglés cuentan como polaridad y subjetividad 0 pero traducirlos mientras se procesan es demasiado costoso.

**Solución empleada:** los tweets que tengan una polaridad/subjetividad de 0 tendrán un peso de 0.25 y los demás tendrán un peso de 1.

A la hora de calcular la media de la poladidad y subjetividad se usará una media ponderada basada en esos pesos asignados.
Al implementar esa solución crearemos dos columnas más para almacenar los pesos de la polaridad y la subjetividad.

In [0]:
sentimentDF = spark.createDataFrame(tweetCountsDF.rdd.map(lambda tweet:(tweet['text'], TextBlob(tweet['text']).polarity, TextBlob(tweet['text']).subjectivity, 1.0 if TextBlob(tweet['text']).polarity != 0 else 0.25, 1.0 if TextBlob(tweet['text']).subjectivity!= 0 else 0.25))).toDF("Tweet","Polarity","Subjectivity","WeightPolarity","WeightSubjectivity")

Creamos una tabla para poder hacer las consultas SQL.

In [0]:
sentimentDF.createOrReplaceTempView("sentiment19")

######Media de la polaridad de los tweets recogidos como json

El valor de la polaridad fluctúa entre -1 y 1  (de -1 a 0 es negativo, 0 es neutro y de 0 a 1 es positivo).

In [0]:
%sql select sum(Polarity)/ sum(WeightPolarity) as average_polarity from sentiment19

average_polarity 0.05602352586530671

Se puede observar que la media de las polaridades roza  la neutralidad incluso tras haber reducido el peso de las polaridades de 0 causadas por los posibles textos en otros idiomas. 

Esto quiere decir que las opiniones sobre el tema están muy reñidas y hay cantidades similares de sentimientos negativos y positivos. Por otro lado, encontramos otro problema a la hora de analizar texto: **la ironía**.

Este modo de sentiment analysis le otorga puntuaciones a ciertas palabras que se encuentran dentro de un diccionario, y luego estas se multiplican dentro de la frase que forma un patrón para generar el resultado final: https://github.com/sloria/TextBlob/blob/dev/textblob/_text.py

Es decir, que si hay varias palabras positivas, es posible que revele un resultado positivo siendo la intención de la persona lo contrario.

######Media de la subjetividad de los tweets recogidos como json

El valor de la subjetividad fluctúa entre 0 (hecho) y 1 (opinión)

In [0]:
%sql select sum(Subjectivity)/sum(WeightSubjectivity) as average_subjectivity from sentiment19

average_subjectivity 0.3472802721305087

En cuanto a la subjetividad, esta se mide contemplando la intensidad del texto, que se resume en el uso de modificadores como adjetivos, etc. Si se usa "Muy mal", "Odioso", "Genial" ....se tomará como una opinión que se aleja de los hechos.

En este caso, ya que no se encuentra tan reñido como el caso anterior se puede decir que se tiende más a los hechos: esto se debe a la temática del análisis plenamente, ya que en muchos casos los tweets son usuarios comentando lo que está pasando en su zona o region, o en general, el mundo como la llegada de las vacunas, datos de contagio etc.
A continuación, veremos algunos tweets que servirán como ejemplo:
* New campaign urges public to get tested twice a week
* Corona Info for #India: New Cases: 131893 Today Recovery: 61836 Today Deaths: 802 Total Active Cases: 979519
* A new Danish study finds no connection between the degree of shutdowns and mortality in European countries evaluated

Podemos observar que apenas usan modificadores y eso afecta en gran medida al resultado del cálculo de la subjetividad.

#####Twitter Sentiment Analysis utilizando Naive Bayes

Se utilizará **Naive Bayes Multinomial** ya que su premisa es bastante buena en comparación con otros modelos en cuanto a clasificación de texto. Se ha consultado el siguiente vídeo sobre Naive Bayes Multinomial y su forma de clasificar texto: https://www.youtube.com/watch?v=O2L2Uv9pdDA.

Al ser naive pasará por alto la interconexión que pueda haber entre las palabras, es decir, todas las palabras tendrán el mismo valor.

Para poder analizarlo correctamente se elimina el ruido y posteriormente se procede a hacer lo siguiente:
el analizador calculará la probabilidad de que sea positivo y la de que sea negativo, partiendo de la probabilidad de que sea negativo/positivo dada por el dataset con el que se ha entrenado el modelo y multiplicado por la probabilidad de que aparezca cada palabra sabiendo que es negativo o positivo.

Es decir, dado:

* N = negativo

* P = positivo

* X1......Xm = las palabras del tweet

Puntuación negativa = p(N) x p(X1/N) .... x p(Xm/N))

Puntuación positiva = p(P) x P(X1/P) .... x P(Xm/P))

La suma de las puntuaciones debe ser 1 y si la puntuación negativa es mayor que la positiva, el sentimiento será negativo y viceversa.

Primero importaremos **nltk (Natural Language Toolkit)** y el dataset que contiene reseñas de películas que se utilizarán para entrenar el modelo.
También importamos el analizador de Naive Bayes: https://textblob.readthedocs.io/en/dev/_modules/textblob/en/sentiments.html

In [0]:
import nltk
nltk.download('movie_reviews')
nltk.download('punkt')
from textblob.sentiments import NaiveBayesAnalyzer

[nltk_data] Downloading package movie_reviews to /root/nltk_data...
[nltk_data] Unzipping corpora/movie_reviews.zip.
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data] Unzipping tokenizers/punkt.zip.

Redefinimos el listener de nuevo para asegurarnos de que todos los tweets están en inglés y guardarlos en una carpeta nueva ya que deseamos utilizarlos para un fin distinto.

In [0]:
class MyStreamListener(tweepy.StreamListener):
    def __init__(self,api=None):
        super(MyStreamListener,self).__init__()
        self.num_tweets=0
    def on_status(self,status):
      if status.lang == 'en':
          tweet= status._json
          aux = json.dumps(tweet)
          print(aux)
          dbutils.fs.put("/FileStore/tables/sentiment/sentiment-"+str(self.num_tweets)+".json", aux)
          self.num_tweets+=1;
          if self.num_tweets<1000 :
              return True
          else:
              return False
    def on_error(self, status):
        print(status)
        if status == 420: return False
        return True

Al igual que se ha visto en la parte 1, si queremos utilizar otra muestra usamos los siguientes comandos:
* Borrado de la carpeta.
* Creación e inicio del stream que obtiene y guarda los json de los tweets.

 **No activarlos** si se desea mantener las mismas muestras pues eso borrará los datos y cargará otros nuevos.

In [0]:
dbutils.fs.rm('/FileStore/tables/sentiment/', True)

Out[31]: True

In [0]:
stream = Stream(auth, MyStreamListener())
stream.filter(track=['covid','covid-19','covid19','coronavirus'])

{"created_at": "Mon Apr 12 21:40:39 +0000 2021", "id": 1381723966465662976, "id_str": "1381723966465662976", "text": "RT @coaccess: \ud83d\udc89COVID-19 Vaccine Drive-Thru Clinic for the Disability Community (Johnson &amp; Johnson): @PASCOhh is hosting a free COVID-19 va\u2026", "source": "<a href=\"https://mobile.twitter.com\" rel=\"nofollow\">Twitter Web App</a>", "truncated": false, "in_reply_to_status_id": null, "in_reply_to_status_id_str": null, "in_reply_to_user_id": null, "in_reply_to_user_id_str": null, "in_reply_to_screen_name": null, "user": {"id": 331244103, "id_str": "331244103", "name": "Colorado Department of Public Health & Environment", "screen_name": "CDPHE", "location": "Colorado", "url": "http://covid19.colorado.gov", "description": "The Colorado Department of Public Health and Environment is dedicated to protecting and improving the health and environment of the people of Colorado.", "translator_type": "none", "protected": false, "verified": true, "followers_count": 47613, "friends_count": 1589, "listed_count": 646, "favourites_count": 3635, "statuses_count": 7499, "created_at": "Thu Jul 07 21:35:53 +0000 2011", "utc_offset": null, "time_zone": null, "geo_enabled": true, "lang": null, "contributors_enabled": false, "is_translator": false, "profile_background_color": "009ADD", "profile_background_image_url": "http://abs.twimg.com/images/themes/theme1/bg.png", "profile_background_image_url_https": "https://abs.twimg.com/images/themes/theme1/bg.png", "profile_background_tile": false, "profile_link_color": "009ADD", "profile_sidebar_border_color": "FFFFFF", "profile_sidebar_fill_color": "DAECF4", "profile_text_color": "663B12", "profile_use_background_image": false, "profile_image_url": "http://pbs.twimg.com/profile_images/1233119894608039936/GdCjRg6j_normal.jpg", "profile_image_url_https": "https://pbs.twimg.com/profile_images/1233119894608039936/GdCjRg6j_normal.jpg", "profile_banner_url": "https://pbs.twimg.com/profile_banners/331244103/1579809424", "default_profile": false, "default_profile_image": false, "following": null, "follow_request_sent": null, "notifications": null}, "geo": null, "coordinates": null, "place": null, "contributors": null, "retweeted_status": {"created_at": "Mon Apr 12 21:28:02 +0000 2021", "id": 1381720792141025280, "id_str": "1381720792141025280", "text": "\ud83d\udc89COVID-19 Vaccine Drive-Thru Clinic for the Disability Community (Johnson &amp; Johnson): @PASCOhh is hosting a free C\u2026 https://t.co/Cw7UjXXmEn", "display_text_range": [0, 140], "source": "<a href=\"https://buffer.com\" rel=\"nofollow\">Buffer</a>", "truncated": true, "in_reply_to_status_id": null, "in_reply_to_status_id_str": null, "in_reply_to_user_id": null, "in_reply_to_user_id_str": null, "in_reply_to_screen_name": null, "user": {"id": 35959223, "id_str": "35959223", "name": "Colorado Access", "screen_name": "coaccess", "location": "Colorado, USA", "url": "http://www.coaccess.com/", "description": "A local, nonprofit Medicaid health plan that has been caring for you and your health since 1994. Serving members throughout the state of Colorado.", "translator_type": "none", "protected": false, "verified": false, "followers_count": 2620, "friends_count": 602, "listed_count": 56, "favourites_count": 963, "statuses_count": 5006, "created_at": "Tue Apr 28 02:33:07 +0000 2009", "utc_offset": null, "time_zone": null, "geo_enabled": true, "lang": null, "contributors_enabled": false, "is_translator": false, "profile_background_color": "96CCED", "profile_background_image_url": "http://abs.twimg.com/images/themes/theme1/bg.png", "profile_background_image_url_https": "https://abs.twimg.com/images/themes/theme1/bg.png", "profile_background_tile": false, "profile_link_color": "AD8EAD", "profile_sidebar_border_color": "FFFFFF", "profile_sidebar_fill_color": "FFFFFF", "profile_text_color": "08406B", "profile_use_background_image": true, "profile_image_url": "http://pbs.twimg.com/profile_images/1323726611355557888/zTCccblp_normal

Creamos el dataframe que representará al json (importante tener la opción **multiline** activa) al igual que en la parte 1:

In [0]:
sentimentInputDF = (
  spark
    .read
    .option("multiline", True)
    .json("/FileStore/tables/sentiment/")
)

Usaremos la clase blobber para generar un solo modelo que se pueda usar en todos los tweets: https://textblob.readthedocs.io/en/dev/advanced_usage.html#sentiment-analyzers (Blobber: A TextBlob Factory)

Si utilizáramos directamente el analyzer, se entrenaría y generaría un modelo por cada tweet, llegando a tomar más de una hora de procesamiento para calcular las estadísticas.

Convertiremos el dataframe en un RDD para hacer un map y guardar el texto junto con el resultado del sentiment analysis pasado por el modelo.

Se volverá a convertir en un dataframe de dos columnas (el texto y el análisis).

In [0]:
NBanalyzer = Blobber(analyzer=NaiveBayesAnalyzer())
sentimentNBrdd = sentimentInputDF.rdd.map(lambda tweet:(tweet['text'], NBanalyzer(tweet['text']).sentiment))

sentimentNBDF = spark.createDataFrame(sentimentNBrdd).toDF("tweet","sentiment")

sentimentNBCountsDF = (
  sentimentNBDF
    .groupBy(
       sentimentNBDF.sentiment
    )
    .count()
)
sentimentNBCountsDF.cache()

Out[23]: DataFrame[sentiment: struct<classification:string,p_pos:double,p_neg:double>, count: bigint]

Podemos mostrar los tweets con sus respectivos análisis de sentimento para futuras referencias si usamos collect()

Volvemos a crear otra tabla para poder hacer consultas SQL.

In [0]:
sentimentNBCountsDF.createOrReplaceTempView("NBsentiment19")

El resultado del análisis mediante Naive Bayes se devuelve a modo de un struct que contiene:
* classification: sentimiento del texto (positivo o negativo)
* p_pos: valor de la positividad del texto
* p_neg: valor de la negatividad del texto

La suma de p_pos y p_neg debe ser 1.

Si p_pos es mayor, classification será positivo (pos). Por otro lado, si p_neg es mayor, classification será negativo (neg).

Utilizaremos el valor de classification pues es la que te devuelve el resultado concreto del análisis. De esta forma observaremos el número de tweets positivos y negativos en la muestra:

In [0]:
%sql select sentiment.classification, sum(count) as total_count from NBsentiment19 group by sentiment.classification

classification total_count pos 691 neg 309

En este último análisis se puede observar que la cantidad de Tweets con sentimiento positivo es más del doble en comparación con los sentimientos negativos.

Hay que tener en cuenta que el gran problema de analizar utilizando el método de Naive Bayes Multinomial es que trata todas las palabras por igual, lo cuál es diferente a cómo nos comunicamos en realidad ya que añadimos distintos énfasis a las palabras o formamos significados alternativos al unirlas con otras.

Algunos ejemplos de posibles fallas:
* 'Hello, I'm a 21 year old trans woman who is currently struggling financially a lot due to not being able to attend work…", sentiment=Row(classification='pos', p_pos=0.6300678104352279, p_neg=0.36993218956477125)) : La clasificación es positiva, pero no poder trabajar y tener problemas financieros no se podría considerar un sentimiento positivo.

* 'Jim died while biking a few hours after posting this. Every death caused by car-centric infrastructure  is heartbreaking. Ev…', sentiment=Row(classification='pos', p_pos=0.5609517954248001, p_neg=0.43904820457520133)): De nuevo, la clasificación es positiva aunque sea por poco.

* 'Variants now account for 70 percent of COVID-19 cases in the city, according to a new report with maps showing wher…', sentiment=Row(classification='pos', p_pos=0.8125299960311315, p_neg=0.1874700039688667)): En este caso también toma que el hecho de que el 70% de los casos de Covid-19 en una ciudad sean variantes como un sentimiento positivo.

En resumen, aunque el modelo es bueno clasificando palabras, no se puede negar que tiene fallas y, aunque una palabra apareza más veces en frases positivas o negativas, ésta puede cambiar de significado totalmente dependiendo de cómo la uses junto a otras. Sirve como una estimación/aproximación bastante buena, pero imperfecta. Por último, hay que mencionar que tampoco es capaz de detectar casos en los que se usa la ironía o el humor.

##Conclusiones

En conclusión, la herramienta de Twitter como medio de análisis en relación con el Covid-19 no llega a tener el potencial esperado. No se ha podido extraer toda la información que se había planeado debido a las políticas de privacidad de Twitter que, por ejemplo, autorizaban las desactivación de la geolocalización.

Por otro lado, se ha podido observar la eficiencia que tiene Twitter a la hora de detectar los idiomas dentro de los tweets, lo cuál nos ha sido de utilidad para poder clasificar los tweets en idiomas y poder así ver cuáles son las comunidades de hablantes qué más parte toman a la hora de publicar acerca del virus. El análisis de los hashtags también ha mostrado la preocupación referente a temas ligados con el Covid-19 como las dosis de ivermectina para paliar la enfermedad o los exámenes de acceso a la universidad presenciales.

Respecto al análisis de sentimiento, se ha comprobado que en ambos casos ha resultado positivo. En mi opinión, parece que el método de clasificación de Naive Bayes es algo más exacto pues tiene reviews reales que han servido para entrenar el modelo en vez de puntuaciones por palabras y frases. No obstante, los dos poseen ciertas fallas, clasificando erróneamente los tweets y demostrando que todavía falta camino por recorrer hasta que se pueda analizar completamente y con precisión el lenguaje natural con su complejidad y todos sus dobles sentidos.

Por último, aunque la herramienta de Twitter no ha sido totalmente conclusiva con respecto al análisis del Covid-19, esta podría servir para otros análisis y usos como estudios de mercado, tendencias u otros que no dependan de las restricciones de los servicios de protección de privacidad de la red social.